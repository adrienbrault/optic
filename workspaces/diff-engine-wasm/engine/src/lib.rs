#![allow(dead_code, unused_imports, unused_variables)]

use wasm_bindgen::prelude::*;

use optic_diff_engine::{HttpInteraction, InteractionDiffResult, SpecEvent, SpecProjection};

#[wasm_bindgen(start)]
pub fn init() {
  wasm_logger::init(wasm_logger::Config::default());
  std::panic::set_hook(Box::new(console_error_panic_hook::hook));
}

#[wasm_bindgen]
pub fn spec_from_events(spec_json: String) -> Result<WasmSpecProjection, JsValue> {
  let spec_events: Vec<SpecEvent> = serde_json::from_str(&spec_json).unwrap();
  let spec_projection = SpecProjection::from(spec_events);

  Ok(WasmSpecProjection::from(spec_projection))
}

#[wasm_bindgen]
pub fn diff_interaction(
  interaction_json: String,
  spec: &WasmSpecProjection,
) -> Result<String, JsValue> {
  let interaction: HttpInteraction = serde_json::from_str(&interaction_json).unwrap();

  let results: Vec<ResultContainer<InteractionDiffResult>> = spec
      .diff_interaction(interaction)
      .into_iter()
      .map(|result| result.into())
      .collect();

  Ok(serde_json::to_string(&results).unwrap())
}
#[wasm_bindgen]
pub fn get_endpoints_projection(spec: WasmSpecProjection) -> Result<String, JsValue> {
  spec.endpoints_projection()
}

#[wasm_bindgen]
pub struct WasmSpecProjection {
  projection: SpecProjection,
}

impl WasmSpecProjection {
  pub fn diff_interaction(&self, interaction: HttpInteraction) -> Vec<InteractionDiffResult> {
    optic_diff_engine::diff_interaction(&self.projection, interaction)
  }

  pub fn endpoints_projection(self) -> Result<String, JsValue> {
    let serializable = self.projection.endpoints_serializable();
    Ok(serde_json::to_string(&serializable).unwrap())
  }
}

impl From<SpecProjection> for WasmSpecProjection {
  fn from(projection: SpecProjection) -> Self {
    Self { projection }
  }
}

#[derive(Debug, serde::Deserialize, serde::Serialize)]
struct ResultContainer<T>(T, String);

impl From<InteractionDiffResult> for ResultContainer<InteractionDiffResult> {
  fn from(result: InteractionDiffResult) -> Self {
    let fingerprint = result.fingerprint();
    Self(result, fingerprint)
  }
}


