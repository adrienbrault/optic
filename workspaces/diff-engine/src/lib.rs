#![allow(dead_code, unused_imports, unused_variables)]

mod commands;
mod events;
mod interactions;
mod learn_shape;
mod projections;
mod protos;
mod queries;
mod shapes;
mod spec;
mod state;

#[cfg(feature = "streams")]
pub mod streams;

pub use commands::{CommandContext, EndpointCommand, RfcCommand, SpecCommand, SpecCommandHandler};
pub use cqrs_core::Aggregate;
pub use events::{
  http_interaction::{HttpInteraction, ArbitraryData, Request, Response, Body},
  RfcEvent, SpecChunkEvent, SpecEvent,
};
pub use interactions::analyze_undocumented_bodies;
pub use interactions::diff as diff_interaction;
pub use interactions::result::{BodyAnalysisLocation, BodyAnalysisResult, InteractionDiffResult};
pub use learn_shape::TrailObservationsResult;
pub use projections::{
  EndpointProjection, ShapeProjection, SpecAssemblerProjection, SpecProjection,
};
pub use protos::shapehash;
pub use queries::endpoint::EndpointQueries;
pub use shapes::diff as diff_shape;
pub use spec::append_batch as append_batch_to_spec;
pub use state::{body::BodyDescriptor, SpecIdGenerator};

pub mod errors {
  pub use super::events::EventLoadingError;

  #[cfg(feature = "streams")]
  pub use super::streams::spec_chunks::{SpecChunkLoaderError, SpecChunkWriterError};
}
