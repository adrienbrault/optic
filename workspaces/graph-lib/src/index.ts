export type NodeId = string;

export enum NodeType {
  Path = 'Path',
  Request = 'Request',
  Response = 'Response',
  Body = 'Body',
}

export type Node = {
  id: NodeId;
} & (
  | {
      type: NodeType.Path;
      data: PathNode;
    }
  | {
      type: NodeType.Request;
      data: RequestNode;
    }
  | {
      type: NodeType.Response;
      data: ResponseNode;
    }
  | {
      type: NodeType.Body;
      data: BodyNode;
    }
);

export type PathNode = {
  absolutePathPattern: string;
  pathId: string;
};
export type RequestNode = {
  requestId: string;
  httpMethod: string;
};
export type ResponseNode = {
  responseId: string;
  httpMethod: string;
  httpStatusCode: number;
};

export type BodyNode = {
  httpContentType: string;
  rootShapeId: string;
};

////////////////////////////////////////////////////////////////////////////////
export interface GraphCommandHandler {
  addNode(node: Node): void;

  addEdge(edge: any, sourceNodeId: NodeId, targetNodeId: NodeId): void;
}

////////////////////////////////////////////////////////////////////////////////

export function mapAppend<K, V>(map: Map<K, V[]>, key: K, value: V) {
  const values = map.get(key) || [];
  values.push(value);
  map.set(key, values);
}

////////////////////////////////////////////////////////////////////////////////

export class GraphIndexer implements GraphCommandHandler {
  readonly nodesById: Map<NodeId, Node>;
  readonly nodesByType: Map<NodeType, Node[]>;
  readonly outboundNeighbors: Map<NodeId, Map<NodeType, Node[]>>;
  readonly inboundNeighbors: Map<NodeId, Map<NodeType, Node[]>>;

  constructor() {
    this.nodesByType = new Map();
    this.nodesById = new Map();
    this.outboundNeighbors = new Map();
    this.inboundNeighbors = new Map();
  }

  addNode(node: Node) {
    if (this.nodesById.has(node.id)) {
      throw new Error(
        `could not add a node with an id that already exists in the graph`
      );
    }
    this.unsafeAddNode(node);
  }

  addEdge(edge: any, sourceNodeId: NodeId, targetNodeId: NodeId) {
    const sourceNode = this.nodesById.get(sourceNodeId);
    if (!sourceNode) {
      throw new Error(`expected ${sourceNodeId} to exist`);
    }

    const targetNode = this.nodesById.get(targetNodeId);
    if (!targetNode) {
      throw new Error(`expected ${targetNodeId} to exist`);
    }

    const outboundNeighbors =
      this.outboundNeighbors.get(sourceNodeId) || new Map();
    mapAppend(outboundNeighbors, targetNode.type, targetNode);
    this.outboundNeighbors.set(sourceNodeId, outboundNeighbors);

    const inboundNeighbors =
      this.inboundNeighbors.get(targetNodeId) || new Map();
    mapAppend(inboundNeighbors, sourceNode.type, sourceNode);
    this.inboundNeighbors.set(targetNodeId, inboundNeighbors);
  }

  unsafeAddNode(node: Node) {
    this.nodesById.set(node.id, node);
    mapAppend(this.nodesByType, node.type, node);
  }
}

// this should be generic so we can do wrap<T> and know what type to expect?
export interface NodeWrapper {
  result: Node;
}

export class RequestNodeWrapper implements NodeWrapper {
  constructor(public result: Node, private queries: GraphQueries) {}

  path(): PathNodeWrapper {
    const neighbors = this.queries.listOutgoingNeighborsByType(
      this.result.id,
      NodeType.Path
    );
    if (neighbors.results.length === 0) {
      throw new Error(`expected Request to have a parent Path`);
    }
    return neighbors.results[0] as PathNodeWrapper;
  }

  bodies(): NodeListWrapper {
    return this.queries.listIncomingNeighborsByType(
      this.result.id,
      NodeType.Body
    );
  }
}

export class ResponseNodeWrapper implements NodeWrapper {
  constructor(public result: Node, private queries: GraphQueries) {}

  path(): PathNodeWrapper {
    const neighbors = this.queries.listOutgoingNeighborsByType(
      this.result.id,
      NodeType.Path
    );
    if (neighbors.results.length === 0) {
      throw new Error(`expected Response to have a parent Path`);
    }
    return neighbors.results[0] as PathNodeWrapper;
  }

  bodies(): NodeListWrapper {
    return this.queries.listIncomingNeighborsByType(
      this.result.id,
      NodeType.Body
    );
  }
}

export class PathNodeWrapper implements NodeWrapper {
  constructor(public result: Node, private queries: GraphQueries) {}

  requests(): NodeListWrapper {
    return this.queries.listIncomingNeighborsByType(
      this.result.id,
      NodeType.Request
    );
  }

  responses(): NodeListWrapper {
    return this.queries.listIncomingNeighborsByType(
      this.result.id,
      NodeType.Response
    );
  }
}

export interface NodeListWrapper {
  results: NodeWrapper[];
}

export class GraphQueries {
  constructor(private index: GraphIndexer) {}

  findById(id: NodeId): NodeWrapper | null {
    const node = this.index.nodesById.get(id);
    if (!node) {
      return null;
    }
    return this.wrap(node);
  }

  listNodesByType(type: NodeType): NodeListWrapper {
    return this.wrapList(type, this.index.nodesByType.get(type) || []);
  }

  //@TODO add singular find* variant
  listIncomingNeighborsByType(id: NodeId, incomingNeighborType: NodeType) {
    const neighbors = this.index.inboundNeighbors.get(id);
    if (!neighbors) {
      return this.wrapList(incomingNeighborType, []);
    }
    const neighborsOfType = neighbors.get(incomingNeighborType);
    return this.wrapList(incomingNeighborType, neighborsOfType || []);
  }

  //@TODO add singular find* variant
  listOutgoingNeighborsByType(
    id: NodeId,
    outgoingNeighborType: NodeType
  ): NodeListWrapper {
    const neighbors = this.index.outboundNeighbors.get(id);
    if (!neighbors) {
      return this.wrapList(outgoingNeighborType, []);
    }
    const neighborsOfType = neighbors.get(outgoingNeighborType);
    return this.wrapList(outgoingNeighborType, neighborsOfType || []);
  }

  wrap(node: Node): NodeWrapper {
    if (node.type === NodeType.Request) {
      return new RequestNodeWrapper(node, this);
    } else if (node.type === NodeType.Response) {
      return new ResponseNodeWrapper(node, this);
    } else if (node.type === NodeType.Path) {
      return new PathNodeWrapper(node, this);
    } else if (node.type === NodeType.Body) {
      return { result: node };
    }
    throw new Error(`unexpected node.type`);
  }

  wrapList(type: NodeType, nodes: Node[]): NodeListWrapper {
    //@TODO add list helpers (map, etc.)
    return {
      results: nodes.map((node) => this.wrap(node)),
    };
  }
}
