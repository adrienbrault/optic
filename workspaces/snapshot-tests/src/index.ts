import path from 'path';
import fs from 'fs-extra';
import { TypedEventEmitter } from './typed-event-emitter';
import {
  executors,
  ITaskContext,
  ITaskIdentifier,
  ITaskSpecification,
  TaskIdentifier,
  TaskType
} from './tasks';
import {
  EventsToRfcStateTaskOutput,
  EventsToRfcStateTaskSpecification
} from './tasks/EventsToRfcState';
import {
  EventsFileToJsTaskOutput,
  EventsFileToJsTaskSpecification
} from './tasks/EventsFileToJsTaskSpecification';
import {
  InteractionsFileToJsTaskOutput,
  InteractionsFileToJsTaskSpecification
} from './tasks/InteractionsFileToJs';
import {
  RfcStateAndInteractionsToDiffsTaskOutput,
  RfcStateAndInteractionsToDiffsTaskSpecification
} from './tasks/RfcStateAndInteractionsToDiffs';
import DataLoader from 'dataloader';

async function* FileSystemInputIterator(
  inputDirectory: string
): AsyncGenerator<ITaskContext> {
  const inputUniverses = await fs.readdir(path.join(inputDirectory, 'events'));
  for (const universe of inputUniverses) {
    console.log({ universe });
    const eventScenarios = await fs.readdir(
      path.join(inputDirectory, 'events', universe)
    );
    console.log({ universe, eventScenarios });

    const interactionsScenarios = await fs.readdir(
      path.join(inputDirectory, 'interactions', universe)
    );
    console.log({ universe, interactionsScenarios });

    for (const eventsScenario of eventScenarios) {
      for (const interactionsScenario of interactionsScenarios) {
        console.log(`emitting next input scenario`);
        const scenario = {
          universe,
          eventsScenario,
          interactionsScenario
        };
        console.log({ scenario });
        yield scenario;
      }
    }
  }
}

import crypto from 'crypto';

function getOutputFilePrefix(id: TaskIdentifier) {
  const hash = crypto.createHash('sha256');
  hash.update(id);
  return hash.digest('hex');
}

function WrapError(e: Error) {
  return {
    type: 'error',
    value: e.message
  };
}

function WrapSuccess(result: any) {
  return {
    type: 'success',
    value: result
  };
}

//@CONTRIBUTORS: this is where we build the dependency graph of tasks
function TaskSpecificationsForContext(
  context: ITaskContext
): ITaskSpecification[] {
  const eventsFileToJs: EventsFileToJsTaskSpecification = {
    type: TaskType.EventsFileToJsTaskSpecification,
    context,
    inputs: {}
  };
  const eventsToRfcState: EventsToRfcStateTaskSpecification = {
    type: TaskType.EventsToRfcStateTaskSpecification,
    context,
    inputs: {
      eventsFromFile: eventsFileToJs
    }
  };
  const interactionsFileToJs: InteractionsFileToJsTaskSpecification = {
    type: TaskType.InteractionsFileToJsTaskSpecification,
    context,
    inputs: {}
  };
  const eventsAndInteractionsToDiffs: RfcStateAndInteractionsToDiffsTaskSpecification = {
    type: TaskType.RfcStateAndInteractionsToDiffs,
    context,
    inputs: {
      interactionsFromFile: interactionsFileToJs,
      eventsFromFile: eventsFileToJs
    }
  };
  return [
    eventsFileToJs,
    interactionsFileToJs,
    eventsToRfcState,
    eventsAndInteractionsToDiffs
  ];
}

////////////////////////////////////////////////////////////////////////////////
//@CONTRIBUTORS register output serializers here
const outputSerializers: {
  [TaskType.EventsToRfcStateTaskSpecification]: (
    output: EventsToRfcStateTaskOutput
  ) => string;
  [TaskType.InteractionsFileToJsTaskSpecification]: (
    output: InteractionsFileToJsTaskOutput
  ) => string;
  [TaskType.EventsFileToJsTaskSpecification]: (
    output: EventsFileToJsTaskOutput
  ) => string;
  [TaskType.RfcStateAndInteractionsToDiffs]: (
    output: RfcStateAndInteractionsToDiffsTaskOutput
  ) => string;
} = {
  [TaskType.EventsFileToJsTaskSpecification](output) {
    return JSON.stringify(output.events, null, 2);
  },
  [TaskType.EventsToRfcStateTaskSpecification](output) {
    return JSON.stringify(output.shapesProjection, null, 2);
  },
  [TaskType.InteractionsFileToJsTaskSpecification](output) {
    return JSON.stringify(output.interactions, null, 2);
  },
  [TaskType.RfcStateAndInteractionsToDiffs](output) {
    return JSON.stringify(
      output.diffs,
      null,
      2
    );
  }
};
////////////////////////////////////////////////////////////////////////////////

const outputBaseDirectory = 'outputs';

async function serializeFailure(
  id: string,
  taskSpecification: ITaskSpecification,
  e: Error
) {
  const groupDirectory = getOutputFilePrefix(
    JSON.stringify(taskSpecification.context)
  );
  const taskDirectory = getOutputFilePrefix(id);
  const outputPath = path.join(
    outputBaseDirectory,
    groupDirectory,
    taskDirectory
  );
  await fs.ensureDir(outputPath);
  await fs.writeJson(
    path.join(
      outputBaseDirectory,
      groupDirectory,
      taskDirectory,
      'specification.json'
    ),
    taskSpecification,
    { spaces: 2 }
  );
  await fs.writeFile(
    path.join(
      outputBaseDirectory,
      groupDirectory,
      taskDirectory,
      'output.error.json'
    ),
    JSON.stringify(WrapError(e))
  );
}

async function serializeResult(
  id: string,
  taskSpecification: ITaskSpecification,
  result: any
) {
  const groupDirectory = getOutputFilePrefix(
    JSON.stringify(taskSpecification.context)
  );
  const taskDirectory = getOutputFilePrefix(id);
  const outputPath = path.join(
    outputBaseDirectory,
    groupDirectory,
    taskDirectory
  );
  await fs.ensureDir(outputPath);
  await fs.writeJson(
    path.join(
      outputBaseDirectory,
      groupDirectory,
      taskDirectory,
      'specification.json'
    ),
    taskSpecification,
    { spaces: 2 }
  );
  await fs.writeFile(
    path.join(
      outputBaseDirectory,
      groupDirectory,
      taskDirectory,
      'output.snapshot.txt'
    ),
    outputSerializers[taskSpecification.type](result)
  );
}

////////////////////////////////////////////////////////////////////////////////
async function main() {
  const taskSpecificationToTaskId: ITaskIdentifier = function(
    taskSpecification: ITaskSpecification
  ): string {
    return JSON.stringify(taskSpecification);
  };

  const inputQueue = new TypedEventEmitter<{
    added: { taskSpecification: ITaskSpecification };
  }>();

  const outputQueues: {
    [key in TaskType]: TypedEventEmitter<{ [key: string]: any }>;
  } = {
    [TaskType.EventsToRfcStateTaskSpecification]: new TypedEventEmitter<{
      [key: string]: EventsToRfcStateTaskOutput;
    }>(),
    [TaskType.InteractionsFileToJsTaskSpecification]: new TypedEventEmitter<{
      [key: string]: InteractionsFileToJsTaskOutput;
    }>(),
    [TaskType.EventsFileToJsTaskSpecification]: new TypedEventEmitter<{
      [key: string]: EventsFileToJsTaskOutput;
    }>(),
    [TaskType.RfcStateAndInteractionsToDiffs]: new TypedEventEmitter<{
      [key: string]: RfcStateAndInteractionsToDiffsTaskOutput;
    }>()
  };

  inputQueue.on('added', async ({ taskSpecification }) => {
    console.log('input added');
    console.log({ taskSpecification });
    const id = taskSpecificationToTaskId(taskSpecification);
    const outputQueue = outputQueues[taskSpecification.type];
    try {
      const dependencyEntries = Object.entries(taskSpecification.inputs).map(
        (entry) => {
          const [key, value] = entry;
          return {
            key,
            value
          };
        }
      );
      console.log(`waiting for dependencies ${id}`);
      console.log(dependencyEntries);
      const dependenciesArray = await results.loadMany(
        dependencyEntries.map((x) => x.value)
      );
      console.log(`done waiting for dependencies ${id}`);

      const dependencies = dependenciesArray.reduce((acc, item, index) => {
        const key = dependencyEntries[index].key;
        acc[key] = item;
        return acc;
      }, {});
      console.log({ dependencies });

      const executor = executors[taskSpecification.type];
      console.log('executor starting');
      //@ts-ignore
      const result = await executor(taskSpecification, dependencies);
      console.log('executor finished');
      console.log({ result });
      outputQueue.emit(id, result);
      serializeResult(id, taskSpecification, result);
    } catch (e) {
      console.log('executor failed');
      console.error(e);
      debugger;
      outputQueue.emit(id, e);
      serializeFailure(id, taskSpecification, e);
    }
  });
  const results: DataLoader<ITaskSpecification, any> = new DataLoader<ITaskSpecification,
    any,
    string>(
    (batch) => {
      console.log({ batch });
      const promises = batch.map((x) => {
        return new Promise<any>((resolve, reject) => {
          const outputQueue = outputQueues[x.type];
          const id = taskSpecificationToTaskId(x);
          console.log(`listening for ${id}`);
          outputQueue.once(id, (result: any) => {
            console.log(`done listening for ${id}`);
            console.log({ id, result });
            resolve(result);
          });
        });
      });
      return Promise.all(promises);
    },
    {
      cacheKeyFn: taskSpecificationToTaskId,
      maxBatchSize: 1
    }
  );
  const inputIterator = FileSystemInputIterator('./inputs');
  for await (const input of inputIterator) {
    const tasks = TaskSpecificationsForContext(input);
    tasks.forEach((taskSpecification) => {
      inputQueue.emit('added', { taskSpecification });
    });
  }
}

main();
