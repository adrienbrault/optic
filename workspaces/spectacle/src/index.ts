import { graphql } from 'graphql';
import { schema } from './graphql/schema';
import { makeExecutableSchema } from '@graphql-tools/schema';
import { shapes, endpoints } from '@useoptic/graph-lib';
import { EventEmitter } from 'events';

export interface IOpticSpecRepository {
  listEvents(): Promise<any[]>;
}

export interface IOpticSpecReadWriteRepository extends IOpticSpecRepository {
  appendEvents(events: any[]): Promise<void>;
  notifications: EventEmitter;
}

export interface IStartDiffResponse {
  diffId: string;
  notificationsUrl?: string;
}

export interface IOpticCaptureService {
  startDiff(
    events: any[],
    ignoreRequests: string[]
  ): Promise<IStartDiffResponse>;
  loadInteraction(pointer: string): Promise<any>;
}

export interface IOpticDiffService {
  listDiffs(): Promise<any[]>;
  listUnrecognizedUrls(): Promise<any[]>;
  learnInitial(events: any[], pathId: string, method: string): Promise<any>;
}

export interface IOpticContext {
  specRepository: IOpticSpecRepository;
  captureService?: IOpticCaptureService;
  diffService?: IOpticDiffService;
}

export interface IBaseSpectacle {
  query(options: SpectacleInput): Promise<any>;
  mutate(options: SpectacleInput): Promise<any>;
}

export interface IForkableSpectacle extends IBaseSpectacle {
  fork(): Promise<IBaseSpectacle>;
}

function buildEndpointsGraph(spec: any, opticEngine: any) {
  const serializedGraph = JSON.parse(
    opticEngine.get_endpoints_projection(spec)
  );
  const { nodes, edges, nodeIndexToId } = serializedGraph;

  const indexer = new endpoints.GraphIndexer();

  function remapId(arrayIndex: number) {
    const fallbackId = arrayIndex.toString();
    const id = nodeIndexToId[fallbackId];
    if (id !== undefined) {
      return id;
    }
    return fallbackId;
  }

  nodes.forEach((node: endpoints.Node, index: number) => {
    const id = remapId(index);
    indexer.addNode({
      ...node,
      id,
    });
  });
  edges.forEach((e: [number, number, any]) => {
    const [sourceIndex, targetIndex, edge] = e;
    indexer.addEdge(edge, remapId(sourceIndex), remapId(targetIndex));
  });
  const queries = new endpoints.GraphQueries(indexer);
  return queries;
}

function buildShapesGraph(spec: any, opticEngine: any) {
  const serializedGraph = JSON.parse(opticEngine.get_shapes_projection(spec));
  const { nodes, edges, nodeIndexToId } = serializedGraph;

  const indexer = new shapes.GraphIndexer();

  function remapId(arrayIndex: number) {
    const fallbackId = arrayIndex.toString();
    const id = nodeIndexToId[fallbackId];
    if (id !== undefined) {
      return id;
    }
    return fallbackId;
  }

  nodes.forEach((node: shapes.Node, index: number) => {
    const id = remapId(index);
    indexer.addNode({
      ...node,
      id,
    });
  });
  edges.forEach((e: [number, number, any]) => {
    const [sourceIndex, targetIndex, edge] = e;
    indexer.addEdge(edge, remapId(sourceIndex), remapId(targetIndex));
  });
  const queries = new shapes.GraphQueries(indexer);
  return queries;
}

type EndpointChanges = {
  data: {
    endpoints: {
      change: {
        category: string;
      };
      path: string;
      method: string;
    }[];
  };
};

function buildEndpointChanges(
  queries: endpoints.GraphQueries,
  since?: string
): EndpointChanges {
  let sortedBatchCommits = queries
    .listNodesByType(endpoints.NodeType.BatchCommit)
    .results.sort((a: any, b: any) => {
      return a.result.data.createdAt < b.result.data.createdAt ? 1 : -1;
    });

  // If there is no `since` date, we want to use every batch commit
  const deltaBatchCommits = since
    ? sortedBatchCommits.filter(
        (batchCommit: any) => batchCommit.result.data.createdAt > since
      )
    : sortedBatchCommits;

  const changes = new Map();

  // Go through requests first. When we go through responses second, we can check
  // to see if the endpoint has been added by a request existing in the batch commits.
  // If the endpoint exists, we can ignore since it's already there. Otherwise we
  // can say the endpoint was updated.
  deltaBatchCommits.forEach((batchCommit: any) => {
    batchCommit.requests().results.forEach((request: any) => {
      const path = request.path().result.data.absolutePathPattern;
      const method = request.result.data.httpMethod;
      const endpointId = JSON.stringify({ path, method });

      // We can always assume a new request means a new endpoint
      changes.set(endpointId, {
        change: {
          category: 'added',
        },
        path,
        method,
      });
    });

    batchCommit.responses().results.forEach((response: any) => {
      const pathNode = response.path();
      const path = pathNode.result.data.absolutePathPattern;
      const method = response.result.data.httpMethod;
      const endpointId = JSON.stringify({ path, method });

      // If the endpoint is there, we should ignore this change
      // We can then assume if the endpoint does not exist, it means
      // this endpoint should be marked as updated.
      if (changes.has(endpointId)) return;

      changes.set(endpointId, {
        change: {
          category: 'updated',
        },
        path,
        method,
      });
    });
  });

  return {
    data: {
      endpoints: Array.from(changes.values()),
    },
  } as EndpointChanges;
}

export async function makeSpectacle(
  opticEngine: any,
  opticContext: IOpticContext
) {
  const events = await opticContext.specRepository.listEvents();
  const spec = opticEngine.spec_from_events(JSON.stringify(events));

  const endpointsQueries = buildEndpointsGraph(spec, opticEngine);
  const shapeViewerProjection = JSON.parse(
    opticEngine.get_shape_viewer_projection(spec)
  );

  const resolvers = {
    Query: {
      requests: (parent: any, args: any, context: any, info: any) => {
        return Promise.resolve(
          context.endpointsQueries.listNodesByType(endpoints.NodeType.Request)
            .results
        );
      },
      shapeChoices: (parent: any, args: any, context: any, info: any) => {
        return Promise.resolve(context.shapeViewerProjection[args.shapeId]);
      },
      endpointChanges: (
        parent: any,
        { since }: { since?: string },
        context: any,
        info: any
      ) => {
        const endpointChanges = buildEndpointChanges(endpointsQueries, since);
        return Promise.resolve(endpointChanges);
      },
      batchCommits: (parent: any, args: any, context: any, info: any) => {
        return Promise.resolve(
          context.endpointsQueries.listNodesByType(
            endpoints.NodeType.BatchCommit
          ).results
        );
      },
    },
    HttpRequest: {
      id: (parent: any) => {
        return Promise.resolve(parent.result.data.requestId);
      },
      pathId: (parent: any) => {
        return Promise.resolve(parent.path().result.data.pathId);
      },
      absolutePathPattern: (parent: any) => {
        return Promise.resolve(parent.path().result.data.absolutePathPattern);
      },
      method: (parent: any) => {
        return Promise.resolve(parent.result.data.httpMethod);
      },
      bodies: (parent: any) => {
        return Promise.resolve(parent.bodies().results);
      },
      responses: (parent: endpoints.RequestNodeWrapper) => {
        return Promise.resolve(parent.path().responses().results);
      },
    },
    HttpResponse: {
      id: (parent: any) => {
        return Promise.resolve(parent.result.data.responseId);
      },
      statusCode: (parent: any) => {
        return Promise.resolve(parent.result.data.httpStatusCode);
      },
      bodies: (parent: any) => {
        return Promise.resolve(parent.bodies().results);
      },
    },
    HttpBody: {
      contentType: (parent: any) => {
        return Promise.resolve(parent.result.data.httpContentType);
      },
      rootShapeId: (parent: any) => {
        return Promise.resolve(parent.result.data.rootShapeId);
      },
    },
    OpticShape: {
      id: (parent: any) => {
        return Promise.resolve(parent.shapeId);
      },
      jsonType: (parent: any) => {
        return Promise.resolve(parent.jsonType);
      },
      asArray: (parent: any) => {
        if (parent.jsonType === 'Array') {
          return Promise.resolve(parent);
        }
      },
      asObject: (parent: any) => {
        if (parent.jsonType === 'Object') {
          return Promise.resolve(parent);
        }
      },
    },
    ArrayMetadata: {
      shapeId: (parent: any) => {
        return Promise.resolve(parent.itemShapeId);
      },
    },
    EndpointChanges: {
      opticUrl: (parent: any) => {
        return Promise.resolve(parent.data.opticUrl);
      },
      endpoints: (parent: any) => {
        return Promise.resolve(parent.data.endpoints);
      },
    },
    EndpointChange: {
      change: (parent: any) => {
        return Promise.resolve(parent.change);
      },
      path: (parent: any) => {
        return Promise.resolve(parent.path);
      },
      method: (parent: any) => {
        return Promise.resolve(parent.method);
      },
    },
    EndpointChangeMetadata: {
      category: (parent: any) => {
        return Promise.resolve(parent.category);
      },
    },
    BatchCommit: {
      createdAt: (parent: any) => {
        return Promise.resolve(parent.result.data.createdAt);
      },
      batchId: (parent: any) => {
        return Promise.resolve(parent.result.data.batchId);
      },
      commitMessage: (parent: any) => {
        return Promise.resolve(parent.result.data.commitMessage);
      },
    },
  };

  const executableSchema = makeExecutableSchema({
    typeDefs: schema,
    resolvers,
  });

  return function (input: SpectacleInput) {
    return graphql({
      schema: executableSchema,
      source: input.query,
      variableValues: input.variables,
      operationName: input.operationName,
      contextValue: {
        opticContext,
        endpointsQueries,
        shapeViewerProjection,
      },
    });
  };
}

export interface SpectacleInput {
  query: string;
  variables: {
    [key: string]: string;
  };
  operationName?: string;
}

/*
InMemorySpectacle and LocalCliSpectacle need to have similar interfaces from the perspective of the ui.
When the applyCommands mutation is completed, the spectacle instance should notify its consumer that the state is stale
thus the spectacle instance should expose an on() event emitter, with a 'change' event that triggers when any internal state changes. Since the mutation can trigger other effects downstream, they should be able to trigger the update as well.
 */

export interface InMemorySpecState {
  events: any[];
}

export class InMemorySpecRepository implements IOpticSpecReadWriteRepository {
  private events: any[] = [];

  constructor(
    public notifications: EventEmitter,
    private initialState: InMemorySpecState
  ) {
    this.events.push(...initialState.events);
  }

  async appendEvents(events: any[]): Promise<void> {
    this.events.push(...events);
    this.notifications.emit('change');
  }

  async listEvents(): Promise<any[]> {
    return this.events;
  }
}
