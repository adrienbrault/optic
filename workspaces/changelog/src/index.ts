import * as DiffEngine from '@useoptic/diff-engine-wasm/engine/build';
import { makeSpectacle } from '@useoptic/spectacle';
import { initial, current } from './data';


export async function generateEndpointChanges(initialEvents: any = [], currentEvents: any): Promise<any> {
  let query;

  // We only need to add a "since" to the query if there are initial events.
  if (initialEvents.length) {
    const initialSpectacle = makeSpectacle(DiffEngine, {
      specEvents: initialEvents
    });

    const batchCommitResults = await initialSpectacle({
      query: `{
        batchCommits {
          createdAt
          batchId
        }
      }`,
      variables: {}
    });

    // TODO: consider making this into a query
    const latestBatchCommit = batchCommitResults.data!.batchCommits!
      .reduce((result: any, batchCommit: any) => {
        return batchCommit.createdAt > result.createdAt ? batchCommit : result;
      });

    query = `{
      endpointChanges(since: "${latestBatchCommit.createdAt}") {
        endpoints {
          change {
            category
          }
          path
          method
        }
      }
    }`;
  } else {
    query = `{
      endpointChanges {
        endpoints {
          change {
            category
          }
          path
          method
        }
      }
    }`;
  }

  const currentSpectacle = makeSpectacle(DiffEngine, {
    specEvents: currentEvents
  });

  return await currentSpectacle({
    query,
    variables: {}
  });
}

// TODO: remove
async function main() {
  const results = await generateEndpointChanges(initial, current);
  console.log(JSON.stringify(results, null, 2));
}

// main();
