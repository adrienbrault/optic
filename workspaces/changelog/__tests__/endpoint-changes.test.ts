import * as fs from 'fs';
import { generateEndpointChanges } from '../src';

function loadSpecs(scenario: string) {
  return {
    initialSpec: JSON.parse(fs.readFileSync(`./__tests__/specs/${scenario}/initial.json`).toString()),
    newSpec: JSON.parse(fs.readFileSync(`./__tests__/specs/${scenario}/new.json`).toString()),
  }
}

const scenarios = fs.readdirSync("./__tests__/specs", { withFileTypes: true })
  .filter(item => item.isDirectory())
  .map(item => item.name);

describe("Generate Endpoint Changes", () => {
  scenarios.forEach(scenario => {
    test(`Check ${scenario} scenario`, async () => {
      const { initialSpec, newSpec } = loadSpecs(scenario);
      const result = await generateEndpointChanges(initialSpec, newSpec);
      expect(result).toMatchSnapshot();
    });
  });
});