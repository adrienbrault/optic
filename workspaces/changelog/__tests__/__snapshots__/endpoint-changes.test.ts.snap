// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate Endpoint Changes Check add-method scenario 1`] = `
Object {
  "data": Object {
    "endpointChanges": Object {
      "endpoints": Array [
        Object {
          "change": Object {
            "category": "added",
          },
          "method": "POST",
          "path": "/user",
        },
      ],
    },
  },
}
`;

exports[`Generate Endpoint Changes Check empty-initial scenario 1`] = `
Object {
  "data": Object {
    "endpointChanges": Object {
      "endpoints": Array [
        Object {
          "change": Object {
            "category": "added",
          },
          "method": "POST",
          "path": "/user",
        },
        Object {
          "change": Object {
            "category": "added",
          },
          "method": "GET",
          "path": "/user",
        },
      ],
    },
  },
}
`;
